server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregatorTests.java

```diff
@@ -42,1236 +42,673 @@
-                aggregation -> aggregation.setNumBuckets(300).field(DATE_FIELD),
-                histogram -> {
-                    List<? extends Histogram.Bucket> buckets = histogram.getBuckets();
-                    assertEquals(120, buckets.size());
-                    for (int i = 0; i < 120; i++) {
-                        Histogram.Bucket bucket = buckets.get(i);
-                        assertEquals(startDate.plusSeconds(i * 5), bucket.getKey());
-                        assertEquals(5, bucket.getDocCount());
-                    }
-                });
-        testSearchAndReduceCase(new MatchAllDocsQuery(), dataset,
-                aggregation -> aggregation.setNumBuckets(100).field(DATE_FIELD),
-                histogram -> {
-                    List<? extends Histogram.Bucket> buckets = histogram.getBuckets();
-                    assertEquals(60, buckets.size());
-                    for (int i = 0; i < 60; i++) {
-                        Histogram.Bucket bucket = buckets.get(i);
-                        assertEquals(startDate.plusSeconds(i * 10), bucket.getKey());
-                        assertEquals(10, bucket.getDocCount());
-                    }
-                });
-        testSearchAndReduceCase(new MatchAllDocsQuery(), dataset,
-                aggregation -> aggregation.setNumBuckets(50).field(DATE_FIELD),
-                histogram -> {
-                    List<? extends Histogram.Bucket> buckets = histogram.getBuckets();
-                    assertEquals(20, buckets.size());
-                    for (int i = 0; i < 20; i++) {
-                        Histogram.Bucket bucket = buckets.get(i);
-                        assertEquals(startDate.plusSeconds(i * 30), bucket.getKey());
-                        assertEquals(30, bucket.getDocCount());
-                    }
-                });
-        testSearchAndReduceCase(new MatchAllDocsQuery(), dataset,
-                aggregation -> aggregation.setNumBuckets(15).field(DATE_FIELD),
-                histogram -> {
-                    List<? extends Histogram.Bucket> buckets = histogram.getBuckets();
-                    assertEquals(10, buckets.size());
-                    for (int i = 0; i < 10; i++) {
-                        Histogram.Bucket bucket = buckets.get(i);
-                        assertEquals(startDate.plusMinutes(i), bucket.getKey());
-                        assertEquals(60, bucket.getDocCount());
-                    }
-                });
+        final Map<Integer, Integer> bucketsToExpectedDocCountMap = new HashMap<>();
+        bucketsToExpectedDocCountMap.put(seed, 1);
+        bucketsToExpectedDocCountMap.put(seed / 2, 5);
+        bucketsToExpectedDocCountMap.put(seed / 6, 10);
+        bucketsToExpectedDocCountMap.put(seed / 12, 30);
+        bucketsToExpectedDocCountMap.put(seed / 40, 60);
+        final Integer randomNumberOfBuckets = randomFrom(bucketsToExpectedDocCountMap.keySet());
```

Comment:
Selecting only one case instead of six cases helps, in addition to choosing a random data point instead for asserting on each data point. (600 * 6 asserts perviously compared to 1 assert now)

Reply:
I see, the random selection of one out of six cases makes sense to me. I wonder if there is a big benefit in randomizing what is `randomLength` now. Making that a fixed 600 datapoint like before shouldn't hurt performance very much if you keep the later code where you only assert the expectedDocCount on one of the buckets. By keeping it a fixed number that is evenly divisible by the denominator in the six cases some of the more complicated `expectedSizeCeil` calculations that you have introduced would not be necessary. I think I'd prefer that for all these kind of tests to make them more readable. Wdyt?

Reply:
> This gave performance improvement of another 1-1.5 sec but I am worried if reducing it too much will comprise the exhaustive nature of these tests

My feeling it to not worry too much about loosing the exhaustiveness of the tests in most cases here. In the end the two things we check in the randomized tests currently are 

1. that we get the expected amount of buckets (with a fixed number of datapoints and reduction factors of 1, 2, 5, etc... where the number of poinst are evently spaced and evenly divided by the factors.
2. the buckets have the right size (the current tests check _all_ buckets, but since they should all be equal sized, its okay to select one randomly

As long as those checks stay valid I think it doesn't hurt to reduce the dataset sizes in the individual test methods a bit if it saves runtime of the test.